File: scala/triangleClassification/src/main/scala/com/noackexpected/triangleClassification/TriangleClassifier.scala
package com.noackexpected.triangleClassification

import com.noackexpected.triangleClassification.TriangleType._

class TriangleClassifier {
  def classify(side1: Int, side2: Int, side3: Int) = {
    def isEquilateral(s1: Int, s2: Int, s3: Int): Boolean = (s2 == s3 && s1 == s2)
    def isIsosceles(s1: Int, s2: Int, s3: Int): Boolean = (s2 == s3 || s1 == s2 || s1 == s3)
    def isValidTriangle(s1: Int, s2: Int, s3: Int): Boolean = (s1 + s2 > s3) && (s1 + s3 > s2) && (s2 + s3 > s1)

    (side1, side2, side3) match {
      case (s1, s2, s3) if !isValidTriangle(s1, s2, s3) => throw new IllegalArgumentException("Not a valid triangle")
      case (s1, s2, s3) if isEquilateral(s1, s2, s3) => Equilateral
      case (s1, s2, s3) if isIsosceles(s1, s2, s3) => Isosceles
      case _ => Scalene
    }
  }
}

File: elixir/triangle_classifier/lib/triangle_classifier.ex
defmodule TriangleClassifier do
  @moduledoc """
  Documentation for `TriangleClassifier`.
  """

  @doc """
  This method accepts three numeric values each representing the length of one side of a triangle. It
  returns a symbol, :scalene, :isosceles, or :equilateral, where the return value corresponds to the
  actual type of the triangle.

  ## Examples

      iex> TriangleClassifier.classify(1, 2, 3)
      :scalene

      iex> TriangleClassifier.classify(1, 1, 3)
      :isosceles

  """
  def classify(side1, side2, side3)

  def classify(side1, side2, side3) when side1 == side2 do
    :isosceles
  end

  def classify(_side1, _side2, _side3) do
    :scalene
  end
end
